### PR TITLE
Writing tutorial for high level overview of atomate2 concepts.

### PR DESCRIPTION
**Pull request by QuantumChemist**
## Summary

This PR will add a tutorial for atomate2, giving a detailed overview of its key concepts.

* Job and flow makers 
* Input sets
* task docs
* builders

## TODO  Checklist

The following sections have to be added

* [x] Job makers
* [x] Flow makers
* [x] Input sets
* [x] Task docs
* [x] Builders
* [x] checking typos
* [x] revise the text

### CODE REVIEW

**Comment by QuantumChemist** [comment]
For now, I have simply drafted a markdown file.

**Comment by QuantumChemist** [comment]
Hey Alex @utf 
If you can spare some time and don't mind, could you have a look at the first section I added, if this is what you expect the tutorial to be? Does it have to be more detailed or shorter? Are the examples ok like that?
Thank you a lot!

**Comment by utf** [comment]
Hi @QuantumChemist, thanks for making a start on this. 

If possible, I would aim to make the tutorial more pedagogical with more emphasis on what the concepts are for rather than delving too much into the code specifics. In general, I imagine that this tutorial will be the first time users are exposed to these concepts, so the main aim is for users to understand why they need to work with these objects and what they achieve, rather than how exactly they work (although this could be included later on in the tutorial).

For example, it could be helpful to have a high level overview at the beginning of the document summarising verbally or graphically how these components interact and form a workflow. You can then give delve into each topic. I would start by giving examples from the atomate2 code and state what is happening in each case. And then if necessary you can give detailed insights into the jobflow source. Although I think this should be minimised as much as possible.

In terms of the writing style and level of detail, I would use the documentation of [QuAcc as an example](https://quantum-accelerators.github.io/quacc/user/recipes/recipes_intro.html).

**Comment by QuantumChemist** [comment]
> Hi @QuantumChemist, thanks for making a start on this.
> 
> If possible, I would aim to make the tutorial more pedagogical with more emphasis on what the concepts are for rather than delving too much into the code specifics. In general, I imagine that this tutorial will be the first time users are exposed to these concepts, so the main aim is for users to understand why they need to work with these objects and what they achieve, rather than how exactly they work (although this could be included later on in the tutorial).
> 
> For example, it could be helpful to have a high level overview at the beginning of the document summarising verbally or graphically how these components interact and form a workflow. You can then give delve into each topic. I would start by giving examples from the atomate2 code and state what is happening in each case. And then if necessary you can give detailed insights into the jobflow source. Although I think this should be minimised as much as possible.
> 
> In terms of the writing style and level of detail, I would use the documentation of [QuAcc as an example](https://quantum-accelerators.github.io/quacc/user/recipes/recipes_intro.html).

Ok, thank you so much for the feedback! :)

**Comment by QuantumChemist** [comment]
Hi @utf , 

I have now adjusted the tutorial text and tried to make it more pedagogical. And I'm very sorry, I was misunderstanding the "high-level overview" part a little bit, and upon looking up the definition, it became clearer to me how the tutorial is supposed to be. I also included a (draft) figure for checking out how that would look like in the tutorial and kept some technical details, as having those helped me a lot in the beginning when I started to work with atomate2. If the current style is ok like that, then I would go on and write the other sections in a similar way and also make proper figures (the one figure is there for demonstration and drafting purposes).  If this is not ok, I'd be very happy about more feedback on my text.

**Comment by utf** [comment]
Hi @QuantumChemist, this looks perfect. Thank you.

File: docs/user/key_concepts_overview.md [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    ) -> LobsterTaskDocument:
+        """
+        Run a LOBSTER calculation.
+        [...]
+        """
+```
+This class incorporates [LOBSTER](http://cohp.de/) specific input and output data, i.e. the `wavefunction_dir` and `basis_dict` as input in `make` that returns the `LobsterMaker`-class specific output as a `TaskDocument`.
+As a job maker, this maker will then create jobs to execute the LOBSTER runs and store the output in the `LobsterTaskDocument` format.
+
+In contrast to a job maker, the `make` function of a `Flow Maker` will return a `Flow` object instead of a task document.
```
Comment (JaGeo) [inline docs/user/key_concepts_overview.md:72]: Maybe highlight the decorator as well 😀

**Comment by QuantumChemist** [comment]
> Hi @QuantumChemist, this looks perfect. Thank you.

Great! Then I will continue the rest of the tutorial like that! Thx!

File: docs/user/key_concepts_overview.md (hunk 1 above)
Comment (QuantumChemist) [reply docs/user/key_concepts_overview.md:72]: Good idea!

File: docs/user/key_concepts_overview.md [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+## Technical Aspects
+
+In atomate2, the `TaskDocument` objects inherit from the [emmet](https://github.com/materialsproject/emmet/) classes `StructureMetadata` or `MoleculeMetadata`.
+
+```
+class StructureMetadata(EmmetBaseModel):
+    """Mix-in class for structure metadata."""
+```
+They contain the structure or molecule metadata, tailored to the variety of computational software.
```
Comment (QuantumChemist) [inline docs/user/key_concepts_overview.md:192]: TODO: Add the verbatim code

**Comment by QuantumChemist** [comment]
Hey @utf  :) 
I would say I'm finished with writing the tutorial. Could you give me some feedback? Also, I hope it's okay. I added the tutorial text to docs/user as one single file!

File: docs/user/key_concepts_overview.md [hunk 3]
```diff
@@ -0,0 +1,335 @@
...
+To streamline the management of such processes, tools like [jobflow](https://github.com/materialsproject/jobflow) can be used. The two essential building blocks of the jobflow-based workflows in atomate2 are `Jobs` and `Flows`.
+
+### Basics
+
+A `Job` is a single computing job, and potentially can take any python function form, given that their inputs and outputs (return values) can be serialized in a JSON format. Jobs are defined using the `@job` decorator.
+
+A `Flow` is a sequential collection of job or other flow objects. The connectivity and also execution order and dependencies of the different jobs and flow is decided automatically from the job inputs. The output from one job (`job.output`) can be used as the input for the next one. This will therefore establish a connectivity between these two jobs.
+The connectivity between several jobs and flows can be arbitrary, depending on the purpose of the workflow. Flows are usually annotated with the `@dataclass` decorator.
```
Comment (JaGeo) [inline docs/user/key_concepts_overview.md:18]:
```suggestion
The connectivity between several jobs and flows can be arbitrary, depending on the purpose of the workflow.
```
I am a bit confused by the statement on the decorator here.

File: docs/user/key_concepts_overview.md [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+    """
+
+    name: str = "phonon"
+    [...]
+
+    def make(...) -> Flow:
+```
+This maker will return a flow that provides all the necessary steps and subroutines that are needed to complete the phonon calculations.
+Oftentimes, such flows then involve dynamic jobs. For the phonon calculations this means that the number of supercells with individual atomic displacements will be decided upon runtime.
+In this particular case, the flow maker `BasePhononMaker` is also inheriting from `ABC` (Abstract Base Classes). <!--more comprehensive explanation on ABCs?-->
```
Comment (JaGeo) [inline docs/user/key_concepts_overview.md:94]:
```suggestion
In this particular case, the flow maker `BasePhononMaker` is also inheriting from `ABC` (Abstract Base Classes). <!--more comprehensive explanation on ABCs?-->
```
I am actually not even sure the abstract class is still needed as we unified the previous dict names, if I am not mistaken. Likely, https://github.com/materialsproject/atomate2/blob/974b9cddff8edcd1773631283962dd88f163a3ad/src/atomate2/common/flows/phonons.py#L363 could be even removed. Thus, I wouldn't add an explanation on this here.

Comment (JaGeo) [reply docs/user/key_concepts_overview.md:94]: okay, taking back that the last part could be removed. I have just seen one of the newer implementations with yet another way to connect jobs.

File: docs/user/key_concepts_overview.md (hunk 3 above)
Comment (QuantumChemist) [reply docs/user/key_concepts_overview.md:18]: Why are you confused? Do you think it belongs somewhere else in the text? Or do you mean that the @dataclass decorator is not a flow-specific decorated as contrary to the @job decorator?

Comment (JaGeo) [reply docs/user/key_concepts_overview.md:18]: The `@dataclass` is used for the `Maker` but not for the Flow itself, right? 
I was preparing an update on my jobflow lecture and therefore looked at your draft ;). Just to explain why I revisited this.

Comment (QuantumChemist) [reply docs/user/key_concepts_overview.md:18]: Oh yes, maybe I should clarify that it's used for the Maker. Thank you for seeing this :)

Comment (JaGeo) [reply docs/user/key_concepts_overview.md:18]: I would just remove this sentence as you clearly describe it later for the Maker.

**Comment by utf** [comment]
This is fantastic. Thank you @QuantumChemist and @JaGeo.

**Comment by QuantumChemist** [comment]
@utf Thank you as well! 😃

**Comment by QuantumChemist** [comment]
Hey @utf , 
I realized I forgot to link this tutorial somewhere, so so far it's only accessible if one knows the full link https://materialsproject.github.io/atomate2/user/key_concepts_overview.html 

So I wonder: Shall I fix this in another PR? If yes, shall I link it to the User Guide, also in a way that it shows up on the User Guide sidebar on the left or shall I link it to the welcome page? What do you think would be best?

**Pull request merged**